requestmanager/reconciledloader/traversalrecord/traversalrecord.go

```diff
@@ -0,0 +1,181 @@
+package traversalrecord
+
+import (
+	"errors"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-graphsync"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+)
+
+// TraversalRecord records the links traversed by a selector and their paths in a space efficient manner
+type TraversalRecord struct {
+	link          *cid.Cid
+	successful    bool
+	childSegments map[datamodel.PathSegment]int
+	children      []*traversalLink
+}
+
+type traversalLink struct {
+	segment datamodel.PathSegment
+	*TraversalRecord
+}
+
+// NewTraversalRecord returns a new traversal record
+func NewTraversalRecord() *TraversalRecord {
+	return &TraversalRecord{
+		childSegments: make(map[datamodel.PathSegment]int),
+	}
+}
+
+// RecordNextStep records the next step in the traversal into the tree
+// based on its path, link, and whether the load was successful or not
+func (tr *TraversalRecord) RecordNextStep(p []datamodel.PathSegment, link cid.Cid, successful bool) {
+	if len(p) == 0 {
+		tr.link = &link
+		tr.successful = successful
+		return
+	}
+	if _, ok := tr.childSegments[p[0]]; !ok {
+		child := traversalLink{
+			TraversalRecord: NewTraversalRecord(),
+			segment:         p[0],
+		}
+		tr.childSegments[p[0]] = len(tr.children)
+		tr.children = append(tr.children, &child)
+	}
+	tr.children[tr.childSegments[p[0]]].RecordNextStep(p[1:], link, successful)
+}
+
+// AllLinks returns all links traversed for a given record
+func (tr *TraversalRecord) AllLinks() []cid.Cid {
+	if len(tr.children) == 0 {
+		return []cid.Cid{*tr.link}
+	}
+	links := make([]cid.Cid, 0)
+	if tr.link != nil {
+		links = append(links, *tr.link)
+	}
+	for _, v := range tr.children {
+		links = append(links, v.AllLinks()...)
+	}
+	return links
+}
+
+// GetLinks returns all links starting at the path in the tree rooted at 'root'
+func (tr *TraversalRecord) GetLinks(root datamodel.Path) []cid.Cid {
+	segs := root.Segments()
+	switch len(segs) {
+	case 0:
+		if tr.link != nil {
+			return []cid.Cid{*tr.link}
+		}
+		return []cid.Cid{}
+	case 1:
+		// base case 1: get all paths below this child.
+		next := segs[0]
+		if childIndex, ok := tr.childSegments[next]; ok {
+			return tr.children[childIndex].AllLinks()
+		}
+		return []cid.Cid{}
+	default:
+	}
+
+	next := segs[0]
+	if _, ok := tr.childSegments[next]; !ok {
+		// base case 2: not registered sub-path.
+		return []cid.Cid{}
+	}
+	return tr.children[tr.childSegments[next]].GetLinks(datamodel.NewPathNocopy(segs[1:]))
+}
+
+// Verifier allows you to verify series of links loads matches a previous traversal
+// order when those loads are successful
+// At any point it can reconstruct the current path.
+type Verifier struct {
```

Comment:
I mean yea sorta. Technically, we could just re-run the selector against the results, but that seems like A LOT of overkill. 

I get why ipld-prime wants to treat data as a giant abstract structured data source in the sky, but yes block boundaries are a thing and it can be tricky to deal with.